goodai/ltm/memory.py

```diff
@@ -0,0 +1,207 @@
+        return self._sync_call("set_state", state_text=state_text)
+
+    def add_content(
+        self, content: str, timestamp: float = None, keywords: list[str] = None,
+        **metadata: Any,
+    ):
+        self.add_queue.put(
+            dict(content=content, timestamp=timestamp, keywords=keywords) | metadata
+        )
+
+    def retrieve(
+        self, query: str, k: int, max_distance: float = None,
+    ) -> list[RetrievedMemory]:
+        return self._sync_call("retrieve", query=query, k=k, max_distance=max_distance)
+
+    def retrieve_from_keywords(self, keywords: list[str]) -> list[RetrievedMemory]:
+        assert len(keywords) > 0
+        return self._sync_call("retrieve_from_keywords", keywords=keywords)
+
+    def __end__(self):
+        self.mem_server.terminate()
+
+
+class LTMSystem:
+
+    def __init__(
+        self, chunk_capacity: int = 50, chunk_overlap_fraction=0, **other_params,
+    ):
+        self.semantic_memory = AutoTextMemory.create(config=TextMemoryConfig(
+            chunk_capacity=chunk_capacity,
+            chunk_overlap_fraction=chunk_overlap_fraction,
+            **other_params,
+        ))
+        self.keyword_index = defaultdict(list)
+
+    def is_empty(self) -> bool:
+        return self.semantic_memory.is_empty()
+
+    def clear(self):
+        self.semantic_memory.clear()
+
+    def state_as_text(self) -> str:
+        return json.dumps(dict(
+            semantic_memory=self.semantic_memory.state_as_text(),
+            keyword_index=self.keyword_index,
+        ))
+
+    def set_state(self, state_text: str):
+        state = json.loads(state_text)
+        self.keyword_index = defaultdict(list, state["keyword_index"])
+        self.semantic_memory.set_state(state["semantic_memory"])
+
+    def add_content(
+        self, content: str, timestamp: float = None, keywords: list[str] = None,
+        **metadata: Any,
+    ) -> int:
+        keywords = keywords or []
+        content, metadata = self.content_addition_preprocessing(content, keywords, **metadata)
+        text_key = self.semantic_memory.add_text(
+            content, timestamp=timestamp, metadata=metadata,
+        )
+        self.semantic_memory.add_separator()
+        for kw in keywords:
+            self.keyword_index[kw].append(text_key)
+        return text_key
+
+    def retrieve(
+        self, query: str, k: int, max_distance: float = None,
+    ) -> list[RetrievedMemory]:
+        memories = self.semantic_memory.retrieve(query, k)
+        if max_distance is not None:
+            memories = [m for m in memories if m.distance <= max_distance]
+        return memories
+
+    def retrieve_from_keywords(self, keywords: list[str]) -> list[RetrievedMemory]:
+        assert len(keywords) > 0
+        text_keys = set()
+        for kw in keywords:
+            text_keys.update(self.keyword_index[kw])
+        mem: DefaultTextMemory = self.semantic_memory
+        memories = list()
+        for chunk in mem.get_all_chunks():
```

Comment:
Maybe add an early stopping condition on the size of `text_keys`?

Reply:
And is there not a better way to index the underlying structure than just iterating though all of the chunks? Does the text key not indicate where in the structure the data is?

Reply:
I'm afraid not 😞 I did spend a significant amount of time trying to make it as efficient as I could, but I can't even keep an index `text_key -> chunks` because the chunks might change and mess it all up... so yeah, this is the solution we have for now. It's still much faster than the semantic retrieval and iteration though.